### ISSUE TITLE
The CMake Build documentation is too detached from reality

### ISSUE THREAD

**Issue by Mq-b** [google/googletest#4681]
### Does the feature exist in the most recent commit?

No.

### Why do we need this feature?

I don't think anyone built and imported the GTest testing framework the way the documentation did:

```cmake
cmake_minimum_required(VERSION 3.14)
project(my_project)

# GoogleTest requires at least C++14
set(CMAKE_CXX_STANDARD 14)
set(CMAKE_CXX_STANDARD_REQUIRED ON)

include(FetchContent)
FetchContent_Declare(
  googletest
  URL https://github.com/google/googletest/archive/03597a01ee50ed33e9dfd640b249b4be3799d395.zip
)
# For Windows: Prevent overriding the parent project's compiler/linker settings
set(gtest_force_shared_crt ON CACHE BOOL "" FORCE)
FetchContent_MakeAvailable(googletest)
```

### Describe the proposal.

I find this confusing, in fact I think what we should be writing is building from source and importing it using CMake:

```cmake
find_package(GTest REQUIRED)
target_link_libraries(${PROJECT_NAME} PRIVATE GTest::GTest)
```

Also, I think "gtest_force_shared_crt" should be defined during the configuration step (e.g., `cmake .. -Dgtest_force_shared_crt=ON`), when generating the build system, rather than during the build or usage phase.

Let me repeat, I think our documentation should be written to build from source, like:

```shell
git clone https://github.com/google/googletest
cd googletest
mkdir build
cd build
cmake .. -Dgtest_force_shared_crt=ON
cmake --build . -j
cmake --install .
```

If it's on Windows, it is also necessary to generate the Release version and install it.

### Is the feature specific to an operating system, compiler, or build system version?

No.

**Comment by phetdam** [comment]
I vaguely remember the Google Test CMake quickstart not having the `FetchContent` stuff a while ago (1.10.0 was the latest release at the time). At that time, I think `find_package` was in the CMake quickstart, but I checked the `quickstart-cmake.md` [commit history](https://github.com/google/googletest/commits/main/docs/quickstart-cmake.md) and can't find any historical record. Not sure if this is due to repo merging with Google Mock.

But as you mentioned, on Windows it's common to want both the Debug and Release configs, both of them linking to their respective *shared* [C runtime libraries](https://learn.microsoft.com/en-us/cpp/build/reference/md-mt-ld-use-run-time-library?view=msvc-170). To have Debug and Release libraries in the same installation root, the debug libs need a `d` or other suffix; as noted in #3843 this was removed (probably for the benefit of single-config CMake generators).

Until the documentation is updated this is how I build on Windows (64-bit, shared C runtimes, `d` suffix for debug libs). Assuming Google Test repo has already been cloned and checked out to a specific commit/tag, from top-level:

```cmd
cmake -S . -B build_win -A x64 -Dgtest_force_shared_crt=ON -DCMAKE_DEBUG_POSTFIX=d
cmake --build build_win --config Debug -j
cmake --build build_win --config Release -j
```

Both configs can then be installed into whichever install root is preferred, e.g.

```cmd
cmake --install build_win --prefix ..\googletest-1.15.0 --config Debug
cmake --install build_win --prefix ..\googletest-1.15.0 --config Release
```

This gives the all the expected Google Test and Google Mock debug and release static libraries.

**Comment by Mq-b** [comment]
Thank you, I have omitted some of my actual builds, it's roughly indeed executing two installs and two builds to ensure that both Debug and Release versions are built and installed.

**Comment by Mq-b** [comment]
I am indeed on a Windows environment, but I believe that whether it's Linux or Windows, we should prioritize using `find_package` and `target_link_libraries` to include.

**Comment by phetdam** [comment]
> I am indeed on a Windows environment, but I believe that whether it's Linux or Windows, we should prioritize using `find_package` and `target_link_libraries` to include.

Yes, this is more meaningful on Linux (or similar) environment where often tagged Google Test versions are installed by the system package manager. In this case, it makes more sense to use `find_package` instead of `FetchContent`.

When I have some time I think I will try to make a PR with some updates to the documentation to highlight these other common methods of Google Test integration with CMake. Otherwise, others might raise similar issues.

**Comment by Mq-b** [comment]
Yes, you have exactly the same idea as me, I also define a `CMAKE_PREFIX_PATH` variable in windows to facilitate the introduction of tripartite libraries. I build all the libraries from source code.

**Comment by phetdam** [comment]
I am going to open a new issue since I am considering adding a new page in the "Guides" section for building from source, as I showed in my [previous comment](https://github.com/google/googletest/issues/4681#issuecomment-2544408031), as well as updating the CMake quickstart page to also include `find_package` usage with separately installed Google Test builds. Will reference this issue to provide additional context.

**Comment by Mizux** [comment]
Personally, I'm using `FetchContent()` everywhere (`google/ortools`, `pybind/pybind_[abseil,protobuf]`, ...)

Also, please note that `FetchContent()` can now nicely interact with `find_package()` thanks to the introduction to
`FIND_PACKAGE_ARGS` and `OVERRIDE_FIND_PACKAGE`

ref: https://cmake.org/cmake/help/latest/module/FetchContent.html#integrating-with-find-package
fun note: They even use googletest as example usage of `FetchContent()`...

**Comment by dkaszews** [comment]
For those looking into using gtest in their project while not necessarily attached to CMake, I highly recommend [xmake](https://xmake.io/) instead as it has very friendly, terse syntax and a built-in package manager with optional support for conan, vcpkg, apt, pacman and many others:

```lua
-- Basic xmake.lua to set up a TDD project
add_requires("gtest >= 1.14.0")

target('my_app')
    set_kind('binary')
    add_files('src/*.cpp')
    add_includedirs('include')

target('my_test')
    set_kind('binary')
    add_files('test/*.cpp')
    add_includedirs('include')
    add_packages('gtest')
```

**Comment by phetdam** [comment]
> Personally, I'm using `FetchContent()` everywhere (`google/ortools`, `pybind/pybind_[abseil,protobuf]`, ...)

Unfortunately there are still legitimate use cases for `find_package`, e.g. in monorepo company environments where dependencies are infrequently updated and already prebuilt. In this case, rebuilding from source for a CI run is extra time spent, as I think most places are not going to follow the live at HEAD policy that Abseil promotes.

I've also had a weird experience building Google's benchmarking library where the Google Test `FetchContent` was overly aggressive in rebuilding the downloaded Google Test sources. Not entirely sure what was going on but it was strange.

> Also, please note that `FetchContent()` can now nicely interact with `find_package()` thanks to the introduction to `FIND_PACKAGE_ARGS` and `OVERRIDE_FIND_PACKAGE`

Thanks for sharing! It's unfortunate, however, that those stuck with pre-3.24 versions of CMake will be missing out.
